GitHub public repo docs (#344)
tinacms/tinacms.org: content/docs/nextjs/github-public-repo.md

---
title: Open Authoring with GitHub
id: /docs/nextjs/github-public-repo
prev: /docs/nextjs/markdown
next:
---

Using Next.js's preview-mode ([announced in Next.js 9.3](https://nextjs.org/blog/next-9-3)), we can load a separate set of data depending on if we are in "edit-mode" or "production-mode". This pairs well with the GitHub API, where we can provide an "Open Authoring" experience where anyone can fork your site, make changes, and create a pull request, all through the TinaCMS UI!

## Requirements:

- You must be using **Next.js v.9.3** or later.
- If your site is static, is must be hosted on a service that supports Next.js preview-mode (E.g [Zeit's Now](https://zeit.co/home))
- Your editable content must be stored in GitHub
- Your repository containing your content **must be public**

## Installation

> We'll need to add a few Tina-GitHub packages to our site:

Reviewer: Do we note the required peerDeps? Assuming this is a fresh install and they may not have `tinacms`, `styled-components` etc.

Reviewer 2: I was thinking the same thing. It would be good if this guide assumed you had nothing setup at all.

> ```

Reviewer: ```suggestion
```bash
```

> npm install --save react-tinacms-github next-tinacms-github

Reviewer: ```suggestion
$ npm install --save react-tinacms-github next-tinacms-github
```

Reviewer 2: I think those `$` should be inserted with CSS or something. Having them in the source messes up the copying

```

or

> ```

Reviewer: ```suggestion
```bash
```

> yarn add react-tinacms-github next-tinacms-github

Reviewer: ```suggestion
$ yarn add react-tinacms-github next-tinacms-github
```

```

#### `react-tinacms-github`

This package provides helpers for setting up TinaCMS to use the GitHub API, with GitHub authentication.

#### `next-tinacms-github`

This package provides helpers for managing the github auth token for requests, as well as
providing helpers for loading content from the GitHub API.

### Environment Variables

Throughout this guide, some environment variables will be used:

You may want a `.env` file in your project root i.e

```
# .env
GITHUB_CLIENT_ID= # We will generate this later when we create a GitHub app.
GITHUB_CLIENT_SECRET= # We will generate this later when we create a GitHub app.
REPO_FULL_NAME=tinacms/tinacms.org # This is your github repository's owner / repo-name.
BASE_BRANCH=master
```

> To load these .env values in the front-end, your `next.config` will need to be configured. We will also use the `dotenv` package to load our `.env` file:

Reviewer: ```suggestion
To load these .env values in the front-end, your [`next.config`](https://nextjs.org/docs/api-reference/next.config.js/introduction) file will need to be configured. We will also use the `dotenv` package to load our `.env` file:
```

> `npm install --save dotenv`

Reviewer: ```suggestion
```bash
$ npm install --save dotenv
```
```

> ```

Reviewer: ```suggestion
```js
```

// next.config.js
require("dotenv").config();

module.exports = {
  env: {
    GITHUB_CLIENT_ID: process.env.GITHUB_CLIENT_ID,
    REPO_FULL_NAME: process.env.REPO_FULL_NAME,
    BASE_BRANCH: process.env.BASE_BRANCH,
  },
  // ...
}
```

> For more help setting up environment variables with Next, see the [Next docs](https://nextjs.org/docs/api-reference/next.config.js/environment-variables)

Reviewer: ```suggestion
For more help setting up environment variables with Next, see the [Next docs](https://nextjs.org/docs/api-reference/next.config.js/environment-variables).
```

> Note that we did not add `GITHUB_CLIENT_SECRET`. This is the a secert key that should only be used from the server and should not be accessible through the browser.

## Setup

> ### Configure the _\_app.tsx_

Reviewer: ```suggestion
### Configure the [custom app](https://nextjs.org/docs/advanced-features/custom-app) file
```

> A few steps must be followed in order to edit your GitHub content using TinaCMS.

Reviewer: ```suggestion
A few steps must be followed in order to edit your GitHub content using TinaCMS. These changes will be made in `_app.tsx`. If you haven't [already created this file](/docs/nextjs/bootstrapping#adding-the-tina-provider), please do so in the `pages` directory.
```

> 1. **Create the TinaCMS**

Reviewer: ```suggestion
1. **Create the TinaCMS instance**
```

1. **Register the GithubClient:** The client accepts a string ('/api/proxy-github' in our case). All requests using the `GithubClient` gets passed through a proxy on our site. This allows us to securely attach the authentication tokens on the backend.
1. Make sure the Sidebar is hidden unless we're in Next's [Preview/Edit mode](https://nextjs.org/docs/advanced-features/preview-mode).
1. **Wrap the Page with `TinacmsGithubProvider`:** This component lets us authenticate with GitHub. It is given config and callbacks that hit our `/api` server functions to enable Preview/Edit Mode after authentication is complete.
1. **Add a button for entering Preview/Edit Mode:** We must provide a means of triggering authentication. This a simple example of how to dow so.

> **pages/\_app.tsx**

Reviewer: ```suggestion
```

Reviewer: To keep the style consistent, we've been adding the example file-paths as a comment in the code block

> ```tsx

Reviewer: ```suggestion
```tsx
// pages/_app.tsx
```

```tsx
// pages/github/authorizing.tsx

import { useGithubAuthRedirect } from 'react-tinacms-github'

// Our GitHub app redirects back to this page with auth code
export default function Authorizing() {
  // Let the main app know, that we received an auth code from the GitHub redirect
  useGithubAuthRedirect()

  return <h2>Authorizing with GitHub, Please wait...</h2>
}
```

## Setup the GitHub OAuth App

In GitHub, within your Account Settings, click <a href="https://github.com/settings/developers" target="_blank">Oauth Apps</a> under Developer Settings. Go ahead and create a "New Oauth App".

For the **Authorization callback URL**, enter the url for the "/github/authorizing" page that you created above (e.g https://your-url/github/authorizing). Fill out the other fields with your custom values.

> _Note:_ If you are **testing your app locally**, you may need a separate development GitHub app (with a localhost redirect), and a production GitHub app.

Next we will add the **Client ID** and **Client Secret** to your environment variables.

<!-- TODO: add ## Developing Locally section at this point? to explain where the content is sourcing from, etc. -->

> ### Setting Environment Variables

Reviewer 2: After this step it's possible to authenticate. It might be nice to point this out and get them to do so.

To set these variables, create a `.env` file in your project root:

```
# .env
GITHUB_CLIENT_ID= # Taken from GitHub
GITHUB_CLIENT_SECRET= # Taken from Github
REPO_FULL_NAME=tinacms/tinacms.org # This is your github repository's owner / repo-name.
BASE_BRANCH=master
```

You can use the `dotenv` package to load the `.env` file:

```bash
$ npm install --save dotenv
```

Now, to load these `.env` values in the front-end, your [next.config](https://nextjs.org/docs/api-reference/next.config.js/introduction) file will need to be configured.

```js
// next.config.js

require('dotenv').config()

module.exports = {
  env: {
    GITHUB_CLIENT_ID: process.env.GITHUB_CLIENT_ID,
    REPO_FULL_NAME: process.env.REPO_FULL_NAME,
    BASE_BRANCH: process.env.BASE_BRANCH,
  },
  // ...
}
```

> Note that we did not add `GITHUB_CLIENT_SECRET` to the config exports. This is the a secret key that should only be used from the server and should not be accessible through the browser.

## Loading content from GitHub

> Now that we have access to the user's auth token, we can load content from GitHub within [_getStaticProps_](https://nextjs.org/docs/basic-features/data-fetching#getstaticprops-static-generation).

Reviewer 2: We should mention the `../content/home.json`.

If that file doesn't exist they could point it to an existing file.

If they don't have any files yet they should create `content/home.json` and make sure to push it to GItHub

## Installation

We'll need to add a few Tina-Github packages to our site:

```
npm install --save react-tinacms-github next-tinacms-github final-form-submit-listener
```

or

```
yarn add react-tinacms-github next-tinacms-github final-form-submit-listener
```

#### `react-tinacms-github`

This package provides helpers for setting up TinaCMS to use the Github API, with Github authentication.

#### `next-tinacms-github`

This package provides helpers for managing the github auth token for requests, as well as
providing helpers for loading content from the Github API.

### Environment Variables

Throughout this guide, some environment variables will be used:

You may want a `.env` file in your project root i.e

```
# .env
GITHUB_CLIENT_ID= # We will generate this later when we create a Github app.
GITHUB_CLIENT_SECRET= # We will generate this later when we create a Github app.
REPO_FULL_NAME=tinacms/tinacms.org # This is your github repository's owner / repo-name.
BASE_BRANCH=master
```

To load these .env values in the front-end, your `next.config` will need to be configured. We will also use the `dotenv` package to load our `.env` file:

`npm install --save dotenv`

```

> require("dotenv").config();

Reviewer: ```suggestion
// next.config.js
require("dotenv").config();
```

module.exports = {
  env: {
    GITHUB_CLIENT_ID: process.env.GITHUB_CLIENT_ID,
    REPO_FULL_NAME: process.env.REPO_FULL_NAME,
    BASE_BRANCH: process.env.BASE_BRANCH,
  },
  // ...
}
```

For help setting up environment variables with Next, see the [Next docs](https://nextjs.org/docs/api-reference/next.config.js/environment-variables)

## Setup

### Register the GithubClient

We will want to use the GithubClient to load/save our content using the Github API. Let's add it as an API plugin.

```ts

> import { TinaCMS } from 'tinacms'

Reviewer: ```suggestion
// pages/_app.js
import { TinaCMS } from 'tinacms'
```

Reviewer: or `.ts`

---
title: Open Authoring with GitHub
id: /docs/nextjs/github-public-repo
prev: /docs/nextjs/markdown
next:
---

Using Next.js's preview-mode ([announced in Next.js 9.3](https://nextjs.org/blog/next-9-3)), we can load a separate set of data depending on if we are in "edit-mode" or "production-mode". This pairs well with the GitHub API, where we can provide an "Open Authoring" experience where anyone can fork your site, make changes, and create a pull request, all through the TinaCMS UI!

## Requirements:

> - You must be using Next.js v.9.3 or later.

Reviewer: ```suggestion
- You must be using **Next.js v.9.3** or later.
```

> - If your site is static, is must be hosted on a service that supports Next.js preview-mode (E.g Zeit's Now)

Reviewer: ```suggestion
- If your site is static, is must be hosted on a service that supports Next.js preview-mode (E.g [Zeit's Now](https://zeit.co/home))
```

- Your editable content must be stored in GitHub

> - Your repository containing your content must be public

Reviewer: ```suggestion
- Your repository containing your content **must be public**
```

> 

Reviewer: We need to add a `tip` (aka blockquote) somewhere to note that this guide works best when your site doesn't have Tina config. Once it's made we can then link to the other Git --> Github migration guide.

```tsx
//pages/github/authorizing.tsx
// Our GitHub app redirects back to this page with auth code
import React from 'react'
import { useGitHubAuthRedirect } from 'react-tinacms-github'

export default function Authorizing() {
  // Let the main app know, that we receieved an auth code from the GitHub redirect
  useGitHubAuthRedirect()
  return <h2>Authorizing with GitHub, Please wait...</h2>
}
```

## GitHub Oauth App

In GitHub, within your account Settings, click [Oauth Apps](https://github.com/settings/developers) under Developer Settings.

click "New Oauth App".

For the **Authorization callback URL**, enter the url for the "authorizing" page that you created above (e.g https://your-url/github/authorizing). Fill out the other fields with your custom values.
_Note: If you are testing your app locally, you may need a separate development GitHub app (with a localhost redirect), and a production GitHub app._

Don't forget to add the **Client ID** and **Client Secret** to your environment variables.

## Loading content from GitHub

Now that we have access to the user's auth token, we can load content from GitHub within `getStaticProps`.

> ```ts

Reviewer: ```suggestion
```tsx
```

//About template about.tsx

> import { getGitHubPreviewProps, GitHubPreviewProps } from 'next-tinacms-github'

Reviewer: ```suggestion
import { 
    getGitHubPreviewProps, 
    GitHubPreviewProps 
} from 'next-tinacms-github'
```

Reviewer: Small thing but it tripped me up — I was trying to import this from 'react-tinacms-github' and didn't notice the different package name since it's almost out of view in the docs.

Reviewer 2: The import's actually wrong too:
```suggestion
import { getGithubPreviewProps, parseMarkdown } from 'next-tinacms-github';
import { GetStaticProps } from 'next';
```

> [See below](#github-oauth-app) for instructions on creating a GitHub OAuth App to generate these **Client ID** & **Client Secret** variables.

### Managing "edit-mode" state

Add the root `TinacmsGithubProvider` component to our main layout. We will supply it with handlers for authenticating and entering/exiting edit-mode.
In this case, we will hit our `/api` server functions.

```tsx
// YourLayout.ts
import { TinacmsGithubProvider } from 'react-tinacms-github';

const enterEditMode = () => {
  return fetch(`/api/preview`).then(() => {
    window.location.href = window.location.pathname
  })
}

const exitEditMode = () => {
  return fetch(`/api/reset-preview`).then(() => {
    window.location.reload()
  })
}

const YourLayout = ({ editMode, error, children }) => {
  return (
    <TinacmsGithubProvider
      clientId={process.env.GITHUB_CLIENT_ID}
      authCallbackRoute='/api/create-github-access-token'
      editMode={editMode}
      enterEditMode={enterEditMode}
      exitEditMode={exitEditMode}
      error={error}>
      {children}
    </TinacmsGithubProvider>
  )

> }

Author: ```suggestion
}
```
```

---
title: Open Authoring with GitHub
id: /docs/nextjs/github-public-repo
prev: /docs/nextjs/markdown
next:
---

Using Next.js's preview-mode ([announced in Next.js 9.3](https://nextjs.org/blog/next-9-3)), we can load a separate set of data depending on if we are in "edit-mode" or "production-mode". This pairs well with the GitHub API, where we can provide an "Open Authoring" experience where anyone can fork your site, make changes, and create a pull request, all through the TinaCMS UI!

## Requirements:

- You must be using **Next.js v.9.3** or later.
- If your site is static, is must be hosted on a service that supports Next.js preview-mode (E.g [Zeit's Now](https://zeit.co/home))
- Your editable content must be stored in GitHub
- Your repository containing your content **must be public**

## Installation

We'll need to add a few Tina-GitHub packages to our site:

```

> npm install --save react-tinacms-github next-tinacms-github final-form-submit-listener

Reviewer 2: ```suggestion
npm install --save react-tinacms-github next-tinacms-github
```

```

or

```

> yarn add react-tinacms-github next-tinacms-github final-form-submit-listener

Reviewer 2: I added this as a dep 
```suggestion
yarn add react-tinacms-github next-tinacms-github
```

```

#### `react-tinacms-github`

This package provides helpers for setting up TinaCMS to use the GitHub API, with GitHub authentication.

#### `next-tinacms-github`

This package provides helpers for managing the github auth token for requests, as well as
providing helpers for loading content from the GitHub API.

### Environment Variables

Throughout this guide, some environment variables will be used:

You may want a `.env` file in your project root i.e

```
# .env
GITHUB_CLIENT_ID= # We will generate this later when we create a GitHub app.
GITHUB_CLIENT_SECRET= # We will generate this later when we create a GitHub app.
REPO_FULL_NAME=tinacms/tinacms.org # This is your github repository's owner / repo-name.
BASE_BRANCH=master
```

To load these .env values in the front-end, your `next.config` will need to be configured. We will also use the `dotenv` package to load our `.env` file:

`npm install --save dotenv`

```
// next.config.js
require("dotenv").config();

module.exports = {
  env: {
    GITHUB_CLIENT_ID: process.env.GITHUB_CLIENT_ID,
    REPO_FULL_NAME: process.env.REPO_FULL_NAME,
    BASE_BRANCH: process.env.BASE_BRANCH,
  },
  // ...
}
```

> For help setting up environment variables with Next, see the [Next docs](https://nextjs.org/docs/api-reference/next.config.js/environment-variables)

Reviewer 2: ```suggestion
For more help setting up environment variables with Next, see the [Next docs](https://nextjs.org/docs/api-reference/next.config.js/environment-variables)
```

## Setup

### Register the GitHubClient

We will want to use the GitHubClient to load/save our content using the GitHub API. Let's add it as an API plugin.

```ts
// pages/_app.js
import { TinaCMS } from 'tinacms'

> import { GitHubClient } from 'react-tinacms-github'
> 
> const REPO_FULL_NAME = process.env.REPO_FULL_NAME as string // e.g: tinacms/tinacms.org
> 
> const cms = new TinaCMS({
>   apis: {
>     github: new GitHubClient('/api/proxy-github', REPO_FULL_NAME),

Reviewer 2: ```suggestion
import { GithubClient } from 'react-tinacms-github'

const REPO_FULL_NAME = process.env.REPO_FULL_NAME as string // e.g: tinacms/tinacms.org

const cms = new TinaCMS({
  apis: {
    github: new GithubClient('/api/proxy-github', REPO_FULL_NAME),
```

> import { TinaCMS } from 'tinacms'
> import { GitHubClient } from 'react-tinacms-github'
> 
> const REPO_FULL_NAME = process.env.REPO_FULL_NAME as string // e.g: tinacms/tinacms.org
> 
> const cms = new TinaCMS({
>   apis: {
>     github: new GitHubClient('/api/proxy-github', REPO_FULL_NAME),
>   },
>   // ... any other tina config
> })

Reviewer 2: Here's a full example of the **_app.tsx**:
```suggestion
import { TinaCMS, TinaProvider } from 'tinacms';
import {
  GithubClient,
  TinacmsGithubProvider,
} from 'react-tinacms-github';
import App from 'next/app';

const REPO_FULL_NAME = process.env.REPO_FULL_NAME as string; // e.g: tinacms/tinacms.org

export default class Site extends App {
  cms: TinaCMS;

  constructor(props) {
    super(props);
    this.cms = new TinaCMS({
      apis: {
        github: new GithubClient('/api/proxy-github', REPO_FULL_NAME),
      },
      sidebar: {
        hidden: !props.pageProps.preview,
      },
      // ... any other tina config
    });
  }
  render() {
    const { Component, pageProps } = this.props;
    return (
      <TinaProvider cms={this.cms}>
        <Component {...pageProps} />
      </TinaProvider>
    );
  }
}
```

Let's setup up some of these backend API functions.

## API functions

We will need a few API functions to handle GitHub authentication, and Next.js's preview-mode:
With Next.js, any functions in the `pages/api` directory are are mapped to `/api/*` endpoints.

We've created a script to make generating these files easier:

From the console, run:

```
$ npx hygen-add https://github.com/dwalkr/hygen-next-tinacms-github
$ npx hygen next-tinacms-github bootstrap --format ts
```

_Note: if your **/pages** directory in not in the root, you will need to supply a `--dir [subDir]` option._

```
$ npx hygen next-tinacms-github bootstrap --format ts --dir src
```

You'll see that a few API functions have been setup for us.

#### `preview.ts`

Contains API function to enter preview-mode, and set the preview-data with content stored in the cookies.

#### `proxy-github.ts`

Contains API function to attach the user's auth token, and proxy requests to the GitHub API.

#### `create-github-auth-token.ts`

Helper for creating a `createCreateAccessToken` server function.

> > _Note: You may need to configure the `process.env.GITHUB_CLIENT_ID` and `process.env.GITHUB_CLIENT_SECRET` environment variables within this file._ _See [Next's documentation](https://nextjs.org/docs/api-reference/next.config.js/environment-variables) for adding environment variables_

Reviewer 2: We already setup the .env file in an earlier step, so I don't think this is needed
```suggestion
```

> > _Note: You may need to configure the `process.env.GITHUB_CLIENT_ID` and `process.env.GITHUB_CLIENT_SECRET` environment variables within this file._ _See [Next's documentation](https://nextjs.org/docs/api-reference/next.config.js/environment-variables) for adding environment variables_
> 
> > [See below](#github-oauth-app) for instructions on creating a GitHub OAuth App to generate these **Client ID** & **Client Secret** variables.

Reviewer 2: We already setup the .env file in an earlier step, so I don't think this is needed
```suggestion
```

> ## Managing "edit-mode" state
> 
> Next, we'll add the root `TinacmsGitHubProvider` component to our main layout. This is responible for validating that the user has access to the repository, and managing entering/exiting edit-mode.
> When we enter/exit edit-mode, we will hit our `/api` server functions that we created in the previous step.
> 
> ```tsx
> // YourLayout.ts
> import { TinacmsGitHubProvider } from 'react-tinacms-github'
> 
> const enterEditMode = () => {
>   return fetch(`/api/preview`).then(() => {
>     window.location.href = window.location.pathname
>   })
> }
> 
> const exitEditMode = () => {
>   return fetch(`/api/reset-preview`).then(() => {
>     window.location.reload()
>   })
> }
> 
> const YourLayout = ({ error, pageProps, children }) => {
>   return (
>     <TinacmsGitHubProvider
>       clientId={process.env.GITHUB_CLIENT_ID || ''}
>       authCallbackRoute="/api/create-github-access-token"
>       enterEditMode={enterEditMode}
>       exitEditMode={exitEditMode}
>       error={pageProps.error}
>     >
>       {children}
>     </TinacmsGitHubProvider>
>   )
> }
> ```

Reviewer 2: ```suggestion
### Managing "edit-mode" state

Add the root `TinacmsGithubProvider` component to our main layout. We will supply it with handlers for authenticating and entering/exiting edit-mode.
In this case, we will hit our `/api` server functions.

```tsx
// YourLayout.ts
import { TinacmsGithubProvider } from 'react-tinacms-github';

const enterEditMode = () => {
  return fetch(`/api/preview`).then(() => {
    window.location.href = window.location.pathname
  })
}

const exitEditMode = () => {
  return fetch(`/api/reset-preview`).then(() => {
    window.location.reload()
  })
}

const YourLayout = ({ editMode, error, children }) => {
  return (
    <TinacmsGithubProvider
      clientId={process.env.GITHUB_CLIENT_ID}
      authCallbackRoute='/api/create-github-access-token'
      editMode={editMode}
      enterEditMode={enterEditMode}
      exitEditMode={exitEditMode}
      error={error}>
      {children}
    </TinacmsGithubProvider>
  )
}
```
```

## Entering / Exiting "edit-mode"

Next, we will need a way to enter/exit mode from our site. Let's create an "Edit Link" button. Ours will take `isEditing` as a parameter.

_We'll be using Next.js's [preview-mode](https://nextjs.org/docs/advanced-features/preview-mode) to set this `isEditing` value. We'll set that up later_

```tsx
//...EditLink.tsx
import React from 'react'

> import { useGitHubEditing } from 'react-tinacms-github'
> 
> export interface EditLinkProps {
>   isEditing: boolean
> }
> 
> export const EditLink = ({ isEditing }: EditLinkProps) => {
>   const github = useGitHubEditing()

Reviewer 2: ```suggestion
import { useGithubEditing } from 'react-tinacms-github'

export interface EditLinkProps {
  isEditing: boolean
}

export const EditLink = ({ isEditing }: EditLinkProps) => {
  const github = useGithubEditing()
```

  return (
    <button onClick={isEditing ? github.exitEditMode : github.enterEditMode}>
      {isEditing ? 'Exit Edit Mode' : 'Edit This Site'}
    </button>
  )
}
```

## Auth Redirects

We will also a page to redirect the user to while authenticating with GitHub.

```tsx
//pages/github/authorizing.tsx
// Our GitHub app redirects back to this page with auth code

> import React from 'react'
> import { useGitHubAuthRedirect } from 'react-tinacms-github'
> 
> export default function Authorizing() {
>   // Let the main app know, that we receieved an auth code from the GitHub redirect
>   useGitHubAuthRedirect()

Reviewer 2: ```suggestion
import { useGithubAuthRedirect } from 'react-tinacms-github'

export default function Authorizing() {
  // Let the main app know, that we receieved an auth code from the GitHub redirect
  useGithubAuthRedirect()
```

  return <h2>Authorizing with GitHub, Please wait...</h2>
}
```

## GitHub Oauth App

> In GitHub, within your account Settings, click [Oauth Apps](https://github.com/settings/developers) under Developer Settings.

Reviewer 2: Open in new tab
```suggestion
In GitHub, within your account Settings, click <a href="https://github.com/settings/developers" target="_blank">Oauth Apps</a> under Developer Settings.
```

click "New Oauth App".

For the **Authorization callback URL**, enter the url for the "authorizing" page that you created above (e.g https://your-url/github/authorizing). Fill out the other fields with your custom values.

> _Note: If you are testing your app locally, you may need a separate development GitHub app (with a localhost redirect), and a production GitHub app._

Reviewer 2: Turn this into a tooltip?
```suggestion
### Authorizing in Development
If you are testing your app locally, you may need a separate development GitHub app (with a localhost redirect), and a production GitHub app.
```

Don't forget to add the **Client ID** and **Client Secret** to your environment variables.

## Loading content from GitHub

Now that we have access to the user's auth token, we can load content from GitHub within `getStaticProps`.

```ts
//About template about.tsx

import { getGitHubPreviewProps, GitHubPreviewProps } from 'next-tinacms-github'

// ...

> export const getStaticProps: GetStaticProps = async function({
>   preview,
>   previewData,
> }) {
>   const filePath = `content/about.md`
> 
>   if (preview) {
>     return getGitHubPreviewProps({
>       ...previewData,
>       fileRelativePath: filePath,
>       format: 'markdown',
>     })
>   } else {
>     // Get your production content here
>     // when you are not in edit-mode.
>     // This should make format of GitHubPreviewData
>     const file = await readLocalMarkdownFile(filePath)
> 
>     return {
>       props: {
>         sourceProviderConnection: null,
>         editMode: false,
>         file,
>         error: null,
>       },
>     }
>   }
> }

Reviewer 2: Shows a usable  fallback
```suggestion
export const getStaticProps: GetStaticProps = async function({
  preview,
  previewData,
}) {
  if (preview) {
    return getGithubPreviewProps({
      ...previewData,
      fileRelativePath: 'src/content/home.json',
      parse: parseMarkdown
    });
  }
  return {
    props: {
      sourceProvider: null,
      error: null,
      preview,
      file: {
        fileRelativePath: 'src/content/home.json',
        data: (await import('../content/home.json')).default,
      },
    },
  };
}
```

```

## Using GitHub Forms

Any forms that we have on our site can be created with the `useGitHubJsonForm` or `useGitHubMarkdownForm` helpers

```tsx

> function BlogTemplate({
>   file, // content for this page
>   sourceProviderConnection, // repository details
> }) {
>   const formOptions = {
>     label: 'About',
>     fields: [
>       //...
>     ],
>   }
> 
>   // Registers a JSON Tina Form
>   const [data, form] = useGitHubJsonForm(file, formOptions, {
>     branch: sourceProvider?.headBranch || '',
>     forkFullName: sourceProvider?.forkFullName || '',
>     baseRepoFullName: process.env.baseRepoFullName || '',
>   })
> 
>   // ...
> }

Reviewer 2: ```suggestion
function BlogTemplate({ jsonFile }) {
  const formOptions = {
    label: 'Blog Post',
    fields: [],
  }

  // Registers a JSON Tina Form
  const [data, form] = useGithubJsonForm(jsonFile, formOptions)

  // ...
}
```

> ## Using GitHub Forms
> 
> Any forms that we have on our site can be created with the `useGitHubJsonForm` or `useGitHubMarkdownForm` helpers
> 
> ```tsx
> function BlogTemplate({
>   file, // content for this page
>   sourceProviderConnection, // repository details
> }) {
>   const formOptions = {
>     label: 'About',
>     fields: [
>       //...
>     ],
>   }
> 
>   // Registers a JSON Tina Form
>   const [data, form] = useGitHubJsonForm(file, formOptions, {
>     branch: sourceProvider?.headBranch || '',
>     forkFullName: sourceProvider?.forkFullName || '',
>     baseRepoFullName: process.env.baseRepoFullName || '',
>   })
> 
>   // ...
> }
> ```

Reviewer 2: ```suggestion
## Using Github Forms

Any forms that we have on our site can be created with the `useGithubJsonForm` or `useGithubMarkdownForm` helpers

```tsx
function BlogTemplate({ jsonFile }) {
  const formOptions = {
    label: 'Blog Post',
    fields: [],
  }

  // Registers a JSON Tina Form
  const [data, form] = useGithubJsonForm(jsonFile, formOptions)

  // ...
}
```

`useGithubJsonForm` will use the `GithubClient` api that we [registered earlier](#register-the-githubclient).
```

`useGitHubJsonForm` will use the `GitHubClient` api that we [registered earlier](#register-the-githubclient).

> ## Error Handling
> 
> We'll also need to add error handling to our forms, which prompts GitHub-specific action when errors occur (e.g a fork no longer exists).
> 
> ```tsx
> // YourSiteForm.ts
> import { useGitHubErrorListener } from 'react-tinacms-github'
> 
> const YourSiteForm = ({ form, children }) => {
>   useGitHubErrorListener(form)
>   return <FormLayout>{children}</FormLayout>
> }
> ```

Reviewer 2: This happens automatically now

```suggestion
```

> ## Error Handling
> 
> We'll also need to add error handling to our forms, which prompts GitHub-specific action when errors occur (e.g a fork no longer exists).
> 
> ```tsx
> // YourSiteForm.ts
> import { useGitHubErrorListener } from 'react-tinacms-github'
> 
> const YourSiteForm = ({ form, children }) => {
>   useGitHubErrorListener(form)
>   return <FormLayout>{children}</FormLayout>
> }
> ```

Reviewer 2: ```suggestion
```

At this point you should be able to authenticate with GitHub! Try clicking the 'Edit This Site' button again and try to authenticate.

<!-- TODO: add an image here to show auth flow modal steps? -->

If all is successful, you should see another modal prompting you to create a fork, go ahead and do so. If you weren't the owner of this repository, a fork would be made where someone else could commit edits and create a PR for review. Since you're the owner of this repository, a fork won't be created and edits will just go to the `master` branch.

## Loading content from GitHub

Now that we have authentication set up, it's time to wire up some content to edit.

Check out `pages/index.tsx`, right now the content for this page is statically written into the component. Let's create a data file to source this content from. Add a new directory in the root of your project called `content` (or `data`, whichever you prefer) and create a new file called `home.json`. We'll start small by just adding editing a title.

```json

> // content/home.json

Author: This comment will break the json. Should we add a note to remove it to the comment? or maybe move it outside the codeblock?

Reviewer 2: Yeah I think convention is usually to put the filename bolded and outside of the codeblock:

    **content/home.json**
    ```json
    {
      "title": "hello world"
    }
    ```

Then the name doesn't end up in the file when you copy-paste it to your site

Author: PR: https://github.com/tinacms/tinacms.org/pull/373

When this gets merged maybe I'll make a separate PR for the rest of the docs

}
```

> > We are currently working on making it possible to make commits directly on the base repository. Until then, to test editing on your own site you will need to test with a separate Github account.

Reviewer 2: It actually works as is. It just thinks you have a fork and commits to master.
